src/lib/ModelConfig.cc

```diff
@@ -17,640 +22,834 @@ namespace AARCH64Opcode {
-          }
-        }
-        // Expand set of blocking groups to include those that inherit from the
-        // user defined set
-        uint16_t config_index =
-            configFile_["Execution-Units"][i]["Blocking-Groups"].size();
-        while (blockingGroups.size()) {
-          // Determine if there's any inheritance
-          if (arch::aarch64::groupInheritance.find(blockingGroups.front()) !=
-              arch::aarch64::groupInheritance.end()) {
-            std::vector<uint16_t> inheritedGroups =
-                arch::aarch64::groupInheritance.at(blockingGroups.front());
-            for (int k = 0; k < inheritedGroups.size(); k++) {
-              blockingGroups.push(inheritedGroups[k]);
-              configFile_["Execution-Units"][i]["Blocking-Groups"]
-                         [config_index] = inheritedGroups[k];
-              config_index++;
-            }
-          }
-          blockingGroups.pop();
-        }
-      }
+  // Core
+  expectations_.addChild(expectations_.create("Core"));
+
+  expectations_["Core"].addChild(
+      expectations_.create<std::string>("ISA", ISA_));
+  expectations_["Core"]["ISA"].setValueSet(
+      std::vector<std::string>{"AArch64", "rv64"});
+
+  // Early check on [Core][ISA] as its value is needed to inform the
+  // expectations of other config options
+  if (!isDefault) {
+    std::string result = expectations_["Core"]["ISA"].validateConfigNode(
+        configTree_["Core"]["ISA"]);
+    configTree_["Core"]["ISA"] >> ISA_;
+    if (result != "Success") {
+      std::cerr << "[SimEng:ModelConfig] Invalid ISA value of \"" << ISA_
+                << "\" passed in config file due to \"" << result
+                << "\" error. Cannot continue with config validation. Exiting."
+                << std::endl;
+      exit(1);
     }
-    subFields.clear();
+  }
+  createGroupMapping();
+
+  expectations_["Core"].addChild(
+      expectations_.create<std::string>("Simulation-Mode", "emulation"));
+  expectations_["Core"]["Simulation-Mode"].setValueSet(
+      std::vector<std::string>{"emulation", "inorderpipelined", "outoforder"});
+
+  expectations_["Core"].addChild(expectations_.create("Clock-Frequency", 1.f));
+  expectations_["Core"]["Clock-Frequency"].setValueBounds(0.f, 10.f);
+
+  expectations_["Core"].addChild(
+      expectations_.create<uint64_t>("Timer-Frequency", 100));
+  expectations_["Core"]["Timer-Frequency"].setValueBounds<uint64_t>(1,
+                                                                    UINT64_MAX);
+
+  expectations_["Core"].addChild(
+      expectations_.create("Micro-Operations", false));
+  expectations_["Core"]["Micro-Operations"].setValueSet(
+      std::vector{false, true});
+
+  if (ISA_ == "AArch64") {
+    expectations_["Core"].addChild(
+        expectations_.create<uint64_t, true>("Vector-Length", 512));
+    expectations_["Core"]["Vector-Length"].setValueSet(
+        std::vector<uint64_t>{128, 256, 384, 512, 640, 768, 896, 1024, 1152,
```

Comment:
The reasoning for using 64 bits for 16-bit options was that values larger than 65535 could be set as config options and then handled safely with the bound check. We could get some overflow issues when reading through ryml otherwise and then the errored config option is harder to handle.